src/fastpath/fastpath.c

```diff
@@ -99,7 +99,23 @@ void NORETURN fastpath_call(word_t cptr, word_t msgInfo)
     stored_hw_asid.words[0] = 0;
 #endif
 #ifdef CONFIG_ARCH_AARCH64
-    stored_hw_asid.words[0] = cap_vtable_root_get_mappedASID(newVTable);
+    /* Need to test that the ASID is still valid */
+    asid_t asid = cap_vtable_root_get_mappedASID(newVTable);
+    asid_map_t asid_map = findMapForASID(asid);
```

Comment:
do we know how many cycles are added by this findMapForASID in fastpath?

Reply:
There's a few changes here that affect the fastpath, so I'll try and break them down.

### Effect of asid_map_t and extra asid_t check on aarch64 non-hyp

aarch64 non-hyp was missing a check for the ASID validity. This means that if an asid_pool_t containing the target thread's VSpace was revoked and then a fastpath was taken the VSpace of the target thread would be switched to even though the VSpace would not have a valid ASID.  The correct behavior would be for an empty table to be assigned to the TTBR as part of the context switch which is what falling back to the slowpath achieves.

For this change I measured a 1 cycle increase in the published numbers (https://sel4.systems/About/Performance/) for default IPC call and reply on the TX1 and TX2. The program should be doing 2 extra memory reads and 2 extra comparisons. I didn't check any event counters yet though and am assuming that the CPU is just making the right magical optimizations and this isn't adding any new stalls.

Any cost increase should be acceptable here as it's adding a check that should be required.

### Effect of asid_map_t and extra asid_t check on aarch64 hyp

aarch64 non-hyp was also missing an asid validity check and would finally call `getHWASID` as part of the context switch. If the asid wasn't valid here then the kernel may dereference a null pointer. As part of these changes, `getHWASID` was removed from the fastpath and replacing it with the changes in the PR the fastpath ends up a few % faster.  I don't have the numbers written down, but can update before final merge. 

### Effect of asid_map_t and removing ASID from page table entry on aarch32 sabre

These changes aren't part of this PR, but I spent some time evaluating an aarch32 implementation assuming it would be most sensitive to costs of this change.  On the sabre, a memory look up is removed as the hw_asid is taken out of the reserved page table entry, but 2 lookups and a conditional branch are added for the ASID table lookups. Published numbers go from 317 cycles (call)  and 336 cycles (recv) on Sabre to 333 and 350 now.  So 16 and 14 more cycles which is 4-5% overhead.